### ISSUE TITLE
too many MemNeededOnStack

### ISSUE THREAD

**Issue by tbutter** [cretz/asmble#4]
The following code fragment (from the larger attached output) causes too many MemNeededOnStack in the if block:

```
(if i32)
  (get_local 3)
  (i32.load align=2)
(else)
  (get_local 6)
  (i32.load align=2)
  (get_local 0)
  (i32.load align=2)
  (call 21)
(end)
```

The inject output looks correct:

```
[TRACE] Stack diff is 0 for If(type=asmble.ast.Node$Type$Value$I32@50378a4)
[TRACE] Stack diff is 1 for GetLocal(index=3)
[TRACE] Injecting asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8 back 1 stack values
[TRACE] Stack diff is 0 for I32Load(align=2, offset=0)

[TRACE] Stack diff is 0 for asmble.ast.Node$Instr$Else@51891008
[TRACE] Stack diff is 1 for GetLocal(index=6)
[TRACE] Injecting asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8 back 1 stack values
[TRACE] Stack diff is 0 for I32Load(align=2, offset=0)
[TRACE] Stack diff is 1 for GetLocal(index=0)
[TRACE] Injecting asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8 back 1 stack values
[TRACE] Stack diff is 0 for I32Load(align=2, offset=0)
[TRACE] Injecting asmble.compile.jvm.Insn$ThisNeededOnStack@aba625 back 2 stack values
[TRACE] Stack diff is -1 for Call(index=21)
[TRACE] Stack diff is 0 for asmble.ast.Node$Instr$End@5db6b9cd

```
(One MemNeededOnStack in the If Block)

But when applying the nodes there are 2 MemNeededOnStack in that block which results in a different stack for the else block:

```
[DEBUG] Applying insn Node(insn=If(type=asmble.ast.Node$Type$Value$I32@50378a4))
[TRACE] Resulting stack: [TypeRef(asm=I)]
[DEBUG] Applying insn asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;)]
[DEBUG] Applying insn asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=Ljava/nio/ByteBuffer;)]
[DEBUG] Applying insn Node(insn=GetLocal(index=3))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=I)]
[DEBUG] Applying insn Node(insn=I32Load(align=2, offset=0))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=I)]


[DEBUG] Applying insn Node(insn=asmble.ast.Node$Instr$Else@51891008)
[DEBUG] Else block for If(type=asmble.ast.Node$Type$Value$I32@50378a4), orig stack [TypeRef(asm=I)]
[TRACE] Resulting stack: [TypeRef(asm=I)]
[DEBUG] Applying insn asmble.compile.jvm.Insn$ThisNeededOnStack@aba625
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;)]
[DEBUG] Applying insn asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=Ljava/nio/ByteBuffer;)]
[DEBUG] Applying insn Node(insn=GetLocal(index=6))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=I)]
[DEBUG] Applying insn Node(insn=I32Load(align=2, offset=0))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=I)]
[DEBUG] Applying insn asmble.compile.jvm.Insn$MemNeededOnStack@7eecb5b8
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;)]
[DEBUG] Applying insn Node(insn=GetLocal(index=0))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=I)]
[DEBUG] Applying insn Node(insn=I32Load(align=2, offset=0))
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=I), TypeRef(asm=I)]
[DEBUG] Applying insn Node(insn=Call(index=21))
[DEBUG] Applying call to $func21 of type Func(params=[asmble.ast.Node$Type$Value$I32@50378a4, asmble.ast.Node$Type$Value$I32@50378a4], ret=asmble.ast.Node$Type$Value$I32@50378a4) with stack [TypeRef(asm=I), TypeRef(asm=Ltest/class;), TypeRef(asm=I), TypeRef(asm=I)]
[TRACE] Resulting stack: [TypeRef(asm=I), TypeRef(asm=I)]
[DEBUG] Applying insn Node(insn=asmble.ast.Node$Instr$End@5db6b9cd)
[DEBUG] End of block If(type=asmble.ast.Node$Type$Value$I32@50378a4), orig stack [TypeRef(asm=I)], unreachable? false
[ERROR] Error in command 'compile': At block end, expected stack [TypeRef(asm=I), TypeRef(asm=Ljava/nio/ByteBuffer;), TypeRef(asm=I)], got [TypeRef(asm=I), TypeRef(asm=I)]
```
[log.txt](https://github.com/cretz/asmble/files/949583/log.txt)

**Comment by cretz** [comment]
I have reworked some of this because of some other bugs (ref issue #1). Can you confirm this is failing on master as well?

**Comment by tbutter** [comment]
Yes, happens with current master.
command line:
bin/asmble compile -log trace ogvjs-1.4.0/ogv-decoder-audio-opus-wasm.wasm test.class

ogv.js from here: https://github.com/brion/ogv.js/releases/tag/1.4.0

**Comment by cretz** [comment]
OK, I will work on narrowing down the issue and develop a test case.

**Comment by tbutter** [comment]
Unfortunately I have no knowledge about WASM so I can't provide a better test case.

**Comment by tbutter** [comment]
I added some debugging code. It seems one of the MemNeededOnStack is caused by a I32Store which is not within the if block.

The meaning of the output is: the injected instructions and the instruction that caused the injection.

I am unsure if that means the MemNeededOnStack should be inserted before the if of within the if AND the else block.
```
[TRACE] Injecting before If(type=asmble.ast.Node$Type$Value$I32@79207381): [] []
[TRACE] Injecting before GetLocal(index=3): [asmble.compile.jvm.Insn$MemNeededOnStack@57db2b13, asmble.compile.jvm.Insn$MemNeededOnStack@57db2b13] [two param I32Store(align=2, offset=0), single param I32Load(align=2, offset=0)]
[TRACE] Injecting before I32Load(align=2, offset=0): [] []
[TRACE] Injecting before asmble.ast.Node$Instr$Else@f68f0dc: [] []
[TRACE] Injecting before GetLocal(index=6): [asmble.compile.jvm.Insn$ThisNeededOnStack@5a5a729f, asmble.compile.jvm.Insn$MemNeededOnStack@57db2b13] [call Call(index=21), single param I32Load(align=2, offset=0)]
[TRACE] Injecting before I32Load(align=2, offset=0): [] []
[TRACE] Injecting before GetLocal(index=0): [asmble.compile.jvm.Insn$MemNeededOnStack@57db2b13] [single param I32Load(align=2, offset=0)]
[TRACE] Injecting before I32Load(align=2, offset=0): [] []
[TRACE] Injecting before Call(index=21): [] []
[TRACE] Injecting before asmble.ast.Node$Instr$End@20b2475a: [] []
```

**Comment by cretz** [comment]
Yeah, the issue is undoubtedly that it's injected before the block start.

**Comment by tbutter** [comment]
It currently is injected within the block. It think it should be injected before. The injectBeforeLastStackCount should probably treat any block as a single instruction with accumulated stack diffs.

**Comment by tbutter** [comment]
This might be a working fix. I never used kotlin before so its probably not correct:
https://github.com/cretz/asmble/compare/master...tbutter:blockstackmanip

**Comment by tbutter** [comment]
I can confirm that I can compile a large project (libopus via emscripten) with the fix above and a fix for #5 without an Exception during compile. (not sure if the resulting class would be running correctly)

**Comment by cretz** [comment]
OK, I think I have fixed it. What I did as I was walking the stack diffs backwards is I now keep a count of inner blocks we're inside of to make sure I don't inject inside of them. Once I get to the top of the block, I adjust the count according to its result (and its pop if it's an `if`). This should make sure that stack injections for things that happen after blocks won't be placed inside of the block.
